lib/Dialect/FIRRTL/Transforms/CheckInit.cpp

```diff
@@ -0,0 +1,318 @@
+}
+
+static void intersectSetSet(SetSet &dest, const SetSet &src) {
+
+  // Filter dest to get rid of things not in src.
+  SmallVector<Value> toDelete;
+  for (auto &[v, bv] : dest)
+    if (!src.count(v))
+      toDelete.push_back(v);
+  for (auto v : toDelete)
+    dest.erase(v);
+
+  // Intersect anything common.  Everything in dest is in src at this point, but
+  // not vice versa.
+  for (auto &[v, bv] : src)
+    if (dest.count(v))
+      dest[v] &= bv;
+}
+
+namespace {
+class CheckInitPass : public CheckInitBase<CheckInitPass> {
+  struct RegionState {
+    // Values Inited in this state's regions.  To be intersected.
+    SetSet init;
+    // Children to merge into this state
+    SmallVector<Operation *, 4> children;
+    // Definitions from this state
+    SmallVector<Value> dests;
+  };
+
+  bool failed;
+  SmallVector<Operation *> worklist;
+  DenseMap<Region *, RegionState> localInfo;
+
+  void processOp(Operation *op, FieldSource &fieldSource);
+  void reportRegion(Value decl, Type t, size_t fieldID, bool needed,
+                    bool alwaysNeeded, BitVector &vals);
+  void emitInitError(Value decl, size_t fieldID);
+
+public:
+  void runOnOperation() override;
+};
+} // end anonymous namespace
+
+static bool checkBit(BitVector &vals, size_t idx) {
+  if (vals.size() <= idx)
+    return false;
+  return vals[idx];
+}
+
+static std::string nameForField(Type t, size_t fieldID) {
+  LLVM_DEBUG(llvm::errs() << "N: " << t << " @" << fieldID << "\n";);
+  return FIRRTLTypeSwitch<Type, std::string>(t)
+      .Case<BundleType, OpenBundleType>([fieldID](auto bundle) {
+        auto idx = bundle.getIndexForFieldID(fieldID);
+        auto fid = bundle.getFieldID(idx);
+        if (fieldID == fid)
+          return bundle.getElementName(idx).str();
+        return bundle.getElementName(idx).str() + "." +
+               nameForField(bundle.getElementType(idx), fieldID - fid);
+      })
+      .Case<FVectorType, OpenVectorType>([fieldID](auto vec) {
+        auto idx = vec.getIndexForFieldID(fieldID);
+        auto fid = vec.getFieldID(idx);
+        if (fieldID == fid)
+          return "[" + std::to_string(idx) + "]";
+        return "[" + std::to_string(idx) + "]." +
+               nameForField(vec.getElementType(), fieldID - fid);
+      })
+      .Default([](auto) { return "INVALID"; });
+}
+
+static StringRef getPortName(BlockArgument b) {
+  FModuleOp mod = cast<FModuleOp>(b.getOwner()->getParentOp());
+  return mod.getPort(b.getArgNumber()).getName();
+}
+
+static StringRef getMemPortName(Value v) {
+  MemOp mem = cast<MemOp>(v.getDefiningOp());
+  return mem.getPortName(cast<OpResult>(v).getResultNumber());
+}
+
+static StringRef getCMemPortName(Value v) {
+  chirrtl::MemoryPortOp mem = cast<chirrtl::MemoryPortOp>(v.getDefiningOp());
+  return mem.getName();
+}
+
+static StringRef getInstPortName(Value v) {
+  InstanceOp inst = cast<InstanceOp>(v.getDefiningOp());
+  return inst.getPortName(cast<OpResult>(v).getResultNumber());
```

Comment:
We have infrastructure for creating names for values to present to users (e.g., `getFieldName`).

Is this different in some way / can or should that be improved?

If there's a reason to have a pass-specific version of this functionality please note it.